snapcraft/projects.py

```diff
@@ -0,0 +1,374 @@
+    def _validate_desktop_name(cls, name):
+        if not re.match(r"^[A-Za-z0-9. _#:$-]+\\.desktop$", name):
+            raise ValueError(
+                f"{name!r} is not a valid desktop file name (e.g. myapp.desktop)"
+            )
+
+        return name
+
+    @pydantic.validator("bus_name")
+    @classmethod
+    def _validate_bus_name(cls, name):
+        if not re.match(r"^[A-Za-z0-9/. _#:$-]*$", name):
+            raise ValueError(f"{name!r} is not a valid bus name")
+
+        return name
+
+    @pydantic.validator(
+        "start_timeout", "stop_timeout", "watchdog_timeout", "restart_delay"
+    )
+    @classmethod
+    def _validate_time(cls, timeval):
+        if not re.match(r"^[0-9]+(ns|us|ms|s|m)*$", timeval):
+            raise ValueError(f"{timeval!r} is not a valid time value")
+
+        return timeval
+
+
+class Hook(ProjectModel):
+    """Snapcraft project hook definition."""
+
+    command_chain: List[CommandChainStr] = []
+    environment: List[Dict[str, str]] = []
+    plugs: UniqueStrList = []
+    passthrough: Optional[Dict[str, Any]]
+
+
+class Architecture(ProjectModel):
+    """Snapcraft project architecture definition."""
+
+    build_on: Union[str, UniqueStrList]
+    build_to: Optional[Union[str, UniqueStrList]]
+
+
+class Project(ProjectModel):
+    """Snapcraft project definition.
+
+    See https://snapcraft.io/docs/snapcraft-yaml-reference
+
+    XXX: Not implemented in this version
+    - environment (top-level)
+    - frameworks
+    - system-usernames
+    - package-repositories
+    - version-script (deprecated)
+    """
+
+    name: constr(max_length=40)  # type: ignore
+    title: Optional[constr(max_length=40)]  # type: ignore
+    base: Optional[str]
+    build_base: Optional[str]
+    compression: Literal["lzo", "xz"] = "xz"
+    version: Optional[constr(max_length=32, strict=True)]  # type: ignore
+    contact: Optional[Union[str, UniqueStrList]]
+    donation: Optional[Union[str, UniqueStrList]]
+    issues: Optional[Union[str, UniqueStrList]]
+    source_code: Optional[str]  # XXX: should we validate as URL?
+    website: Optional[str]  # XXX: should we validate as URL?
+    summary: constr(max_length=78)  # type: ignore
+    description: str
+    type: Literal["app", "base", "gadget", "kernel", "snapd"] = "app"
+    icon: Optional[str]
+    confinement: Literal["classic", "devmode", "strict"]
+    layout: Optional[Dict[str, Dict[str, Any]]]
+    license: Optional[str]
+    grade: Literal["stable", "devel"]
+    adopt_info: Optional[str]
+    architectures: List[Architecture] = []
+    assumes: UniqueStrList = []
+    hooks: Optional[Dict[str, Hook]]
+    passthrough: Optional[Dict[str, Any]]
+    apps: Optional[Dict[str, App]]
+    plugs: Optional[Dict[str, Dict[str, str]]]  # TODO: add plug name validation
+    slots: Optional[Dict[str, Dict[str, str]]]  # TODO: add slot name validation
+    parts: Dict[str, Any]  # parts are handled by craft-parts
+    epoch: Optional[int]
```

Comment:
this is more than int, https://snapcraft.io/docs/snap-epochs

Reply:
Fixed, now matching legacy validation.